### PR TITLE
feat: Add general IDE info to User-Agent for Tabby clients

### PR DESCRIPTION
**Pull request by Sma1lboy**
This addresses the remaining part of the issue mentioned in #2581.

- Retrieve client info based on the client configuration during client initialization.

Example User-Agent format:

```
Node.js/v20.9.0 Visual-Studio-Code-desktop/1.91.1 TabbyML.vscode-tabby/1.8.0-dev
```

### CODE REVIEW

**Comment by Sma1lboy** [comment]
Should I do another PR for intellij? or I can put them together

**Comment by Sma1lboy** [comment]
Sure, I'll take a look tonight

**Comment by Sma1lboy** [comment]
@icycodes Thanks for your suggestions. I've realized that most of my previous code was focused on creating a parameter for ideInfo that could be transmitted by the client. It wasn't until I read the section of the code about lifecycle that I discovered this process had already been written there. :(
I deleted the unnecessary code and cleaned up the user agent format. By the way, if I want to get the current version of the user agent, do I need to read the package.json? I haven't found any code that stores "packagejson" context.

**Comment by icycodes** [comment]
> if I want to get the current version of the user agent, do I need to read the package.json? I haven't found any code that stores "packagejson" context.

Simply import the "package.json" should be ok. Here is an [example](https://github.com/TabbyML/tabby/blob/main/clients/tabby-agent/src/lsp/Server.ts#L93).

File: clients/tabby-agent/src/TabbyAgent.ts [hunk 1]
```diff
@@ -1010,4 +1018,16 @@ export class TabbyAgent extends EventEmitter implements Agent {
       return null;
     }
   }
+
+  private clientInfoToString(session: Record<string, any> | undefined): string {
+    const nodeInfo = `Node.js/${process.version}`;
```
Comment (icycodes) [inline clients/tabby-agent/src/TabbyAgent.ts:1023]: We need to check `isBrowser` here. In browser environment, use `navigator?.userAgent` instead. Here is an [example](https://github.com/TabbyML/tabby/blob/main/clients/tabby-agent/src/AnonymousUsageLogger.ts#L16).

File: clients/tabby-agent/src/TabbyAgent.ts [hunk 2]
```diff
@@ -699,6 +704,9 @@ export class TabbyAgent extends EventEmitter implements Agent {
           temperature,
         },
         signal: this.createAbortSignal({ signal }),
+        headers: {
+          "User-Agent": this.clientInfoStr,
```
Comment (icycodes) [inline clients/tabby-agent/src/TabbyAgent.ts:708]: I think we can add headers when [create http client](https://github.com/TabbyML/tabby/blob/main/clients/tabby-agent/src/TabbyAgent.ts#L163).

File: clients/tabby-agent/src/TabbyAgent.ts [hunk 3]
```diff
@@ -393,6 +395,7 @@ export class TabbyAgent extends EventEmitter implements Agent {
       }
     }
     await this.anonymousUsageLogger.init({ dataStore: this.dataStore });
+    this.clientInfoStr = this.clientInfoToString(undefined);
```
Comment (icycodes) [inline clients/tabby-agent/src/TabbyAgent.ts:398]:
```suggestion
    this.clientInfoStr = this.clientInfoToString(options.clientProperties?.session);
```
This should combine L398 and L410.

**Review by icycodes — CHANGES_REQUESTED** [review]

File: clients/tabby-agent/src/TabbyAgent.ts (hunk 2 above)
Comment (Sma1lboy) [reply clients/tabby-agent/src/TabbyAgent.ts]: I moved user agent into setupApi func

File: clients/tabby-agent/src/TabbyAgent.ts [hunk 4]
```diff
@@ -70,6 +71,7 @@ export class TabbyAgent extends EventEmitter implements Agent {
   private tryingConnectTimer: ReturnType<typeof setInterval>;
   static readonly submitStatsInterval = 1000 * 60 * 60 * 24; // 24h
   private submitStatsTimer: ReturnType<typeof setInterval>;
+  private clientInfoStr: string = "";
```
Comment (wsxiaoys) [inline clients/tabby-agent/src/TabbyAgent.ts:74]:
```suggestion
  private userAgentHeader: string = "";
```

**Review by icycodes — APPROVED** [review]

**Pull request merged**